doc/decisions/reference_counting.md

```diff
@@ -0,0 +1,39 @@
+# Reference Counting
+
+## Problem
+
+- locking is not reset when ref counting again gets 0 (adding to keyset and
+  pop again) #2202
+- C++ API for KeySet and Key has unexpected differences: also use ref counting
+  for KeySets (also suggested in #1332)
+
+## Constraints
+
+## Assumptions
+
+## Considered Alternatives
+
+- make ref counting thread safe (probably useful for JNI)
+- start with 1 for reference counting and let keyDecRef do keyDel
+
+## Decision
+
+- add second reference counter to Key
+- One counter is for locking references, the other one for general references. A locking reference automatically locks/unlocks the keyname.
+- introduce reference counter for KeySets (for external keyset references)
+- `keyIncRef` (or equivalent) return error in case of reference counter overflow
+- use fixed sized types for reference counters
+- increment/decrement references before/after passing instances to plugins
+
+## Rationale
+
+- Adding a second reference counter to Key and reducing the size of both significantly
+  actually saves memory compared to the previous solution.
```

Comment:
not really, in #3949 is now 16bit proposed, so at best we stay at the same size.

Reply:
The current ref counter is `size_t`, so on most modern (i.e. 64 bit) machines we safe 4 Bytes, assuming both counters will be 16bit. #3949 actually only proposes a size of `KeySet` not for `Key`.